src/storage/concurrency_manager/key_handle.rs

```diff
@@ -0,0 +1,153 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+mod lock_store;
+
+use self::lock_store::LockStore;
+use super::lock_table::LockTable;
+
+use std::{mem, sync::Arc};
+use tokio::sync::{Mutex as AsyncMutex, MutexGuard as AsyncMutexGuard};
+use txn_types::{Key, Lock};
+
+/// An entry in the in-memory table providing functions related to a specific
+/// key.
+pub struct KeyHandle {
+    pub key: Key,
+    table: LockTable,
+    mutex: AsyncMutex<()>,
+    lock_store: LockStore,
+}
+
+impl KeyHandle {
+    pub fn new(key: Key, table: LockTable) -> Self {
+        KeyHandle {
+            key,
+            table,
+            mutex: AsyncMutex::new(()),
+            lock_store: LockStore::new(),
+        }
+    }
+
+    pub async fn lock(self: Arc<Self>) -> KeyHandleGuard {
+        // Safety: `_mutex_guard` is declared after `handle_ref` in `KeyHandleGuard`.
+        // So the mutex guard will be released earlier than the `Arc<KeyHandle>`.
+        // Then we can make sure the mutex guard doesn't point to released memory.
+        let mutex_guard = unsafe { mem::transmute(self.mutex.lock().await) };
+        KeyHandleGuard {
+            handle: self,
+            _mutex_guard: mutex_guard,
+        }
+    }
+
+    pub fn with_lock<T>(&self, f: impl FnOnce(&Option<Lock>) -> T) -> T {
+        self.lock_store.read(f)
+    }
+}
+
+impl Drop for KeyHandle {
+    fn drop(&mut self) {
+        self.table.remove(&self.key);
+    }
+}
+
+/// A `KeyHandle` with its mutex locked.
+pub struct KeyHandleGuard {
+    // It must be declared before `handle_ref` so it will be dropped before
+    // `handle_ref`.
+    _mutex_guard: AsyncMutexGuard<'static, ()>,
+    handle: Arc<KeyHandle>,
```

Comment:
Wrong order?

Reply:
https://doc.rust-lang.org/reference/destructors.html
> The fields of a struct are dropped in declaration order.

Reply:
What the mean of the comments above?
```rust
        // Safety: `_mutex_guard` is declared after `handle_ref` in `KeyHandleGuard`.
        // So the mutex guard will be released earlier than the `Arc<KeyHandle>`.
        // Then we can make sure the mutex guard doesn't point to released memory.
```

Reply:
Ah, sorry. It should be `declared before`...